// hmkit/es/CAPABILITY_IDENTIFIERS.js
var _messages, _messages2, _messages4, _messages5, _messages6, _messages7, _messages8, _messages9, _messages11, _messages12, _messages13, _messages14, _messages15, _messages16, _messages17, _messages18, _messages19, _messages21, _messages24, _messages25, _messages26, _messages27, _messages29, _messages31, _messages32, _messages33, _messages34, _messages35, _messages36, _messages37, _messages38, _messages39, _messages40, _messages41, _messages42, _messages43, _messages44, _messages45, _messages46, _messages47;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/* eslint-disable no-useless-computed-key */

export default {
  door_locks: {
    identifier: [0x00, 0x20],
    messages: (_messages = {}, _defineProperty(_messages, 0x00, 'get_lock_state'), _defineProperty(_messages, 0x01, 'lock_state'), _defineProperty(_messages, 0x12, 'lock_unlock_doors'), _messages)
  },
  trunk: {
    identifier: [0x00, 0x21],
    messages: (_messages2 = {}, _defineProperty(_messages2, 0x00, 'get_trunk_state'), _defineProperty(_messages2, 0x01, 'trunk_state'), _defineProperty(_messages2, 0x12, 'control_trunk'), _messages2)
  },
  wake_up: {
    identifier: [0x00, 0x22],
    messages: _defineProperty({}, 0x02, 'wake_up')
  },
  charging: {
    identifier: [0x00, 0x23],
    messages: (_messages4 = {}, _defineProperty(_messages4, 0x00, 'get_charge_state'), _defineProperty(_messages4, 0x01, 'charge_state'), _defineProperty(_messages4, 0x12, 'start_stop_charging'), _defineProperty(_messages4, 0x13, 'set_charge_limit'), _defineProperty(_messages4, 0x14, 'open_close_charging_port'), _defineProperty(_messages4, 0x15, 'set_charge_mode'), _defineProperty(_messages4, 0x16, 'set_charging_timers'), _defineProperty(_messages4, 0x17, 'set_reduction_of_charging_current_times'), _messages4)
  },
  climate: {
    identifier: [0x00, 0x24],
    messages: (_messages5 = {}, _defineProperty(_messages5, 0x00, 'get_climate_state'), _defineProperty(_messages5, 0x01, 'climate_state'), _defineProperty(_messages5, 0x12, 'change_starting_times'), _defineProperty(_messages5, 0x13, 'start_stop_hvac'), _defineProperty(_messages5, 0x14, 'start_stop_defogging'), _defineProperty(_messages5, 0x15, 'start_stop_defrosting'), _defineProperty(_messages5, 0x16, 'start_stop_ionising'), _defineProperty(_messages5, 0x17, 'set_temperature_settings'), _messages5)
  },
  rooftop_control: {
    identifier: [0x00, 0x25],
    messages: (_messages6 = {}, _defineProperty(_messages6, 0x00, 'get_rooftop_state'), _defineProperty(_messages6, 0x01, 'rooftop_state'), _defineProperty(_messages6, 0x12, 'control_rooftop'), _messages6)
  },
  honk_horn_flash_lights: {
    identifier: [0x00, 0x26],
    messages: (_messages7 = {}, _defineProperty(_messages7, 0x00, 'get_flashers_state'), _defineProperty(_messages7, 0x01, 'flashers_state'), _defineProperty(_messages7, 0x12, 'honk_flash'), _defineProperty(_messages7, 0x13, 'activate_deactivate_emergency_flashers'), _messages7)
  },
  remote_control: {
    identifier: [0x00, 0x27],
    messages: (_messages8 = {}, _defineProperty(_messages8, 0x00, 'get_control_mode'), _defineProperty(_messages8, 0x01, 'control_mode'), _defineProperty(_messages8, 0x04, 'control_command'), _defineProperty(_messages8, 0x12, 'start_stop_control'), _messages8)
  },
  valet_mode: {
    identifier: [0x00, 0x28],
    messages: (_messages9 = {}, _defineProperty(_messages9, 0x00, 'get_valet_mode'), _defineProperty(_messages9, 0x01, 'valet_mode'), _defineProperty(_messages9, 0x12, 'activate_deactivate_valet_mode'), _messages9)
  },
  heart_rate: {
    identifier: [0x00, 0x29],
    messages: _defineProperty({}, 0x12, 'send_heart_rate')
  },
  vehicle_location: {
    identifier: [0x00, 0x30],
    messages: (_messages11 = {}, _defineProperty(_messages11, 0x00, 'get_vehicle_location'), _defineProperty(_messages11, 0x01, 'vehicle_location'), _messages11)
  },
  navi_destination: {
    identifier: [0x00, 0x31],
    messages: (_messages12 = {}, _defineProperty(_messages12, 0x00, 'get_navi_destination'), _defineProperty(_messages12, 0x01, 'navi_destination'), _defineProperty(_messages12, 0x12, 'set_navi_destination'), _messages12)
  },
  diagnostics: {
    identifier: [0x00, 0x33],
    messages: (_messages13 = {}, _defineProperty(_messages13, 0x00, 'get_diagnostics_state'), _defineProperty(_messages13, 0x01, 'diagnostics_state'), _messages13)
  },
  maintenance: {
    identifier: [0x00, 0x34],
    messages: (_messages14 = {}, _defineProperty(_messages14, 0x00, 'get_maintenance_state'), _defineProperty(_messages14, 0x01, 'maintenance_state'), _messages14)
  },
  engine: {
    identifier: [0x00, 0x35],
    messages: (_messages15 = {}, _defineProperty(_messages15, 0x00, 'get_ignition_state'), _defineProperty(_messages15, 0x01, 'ignition_state'), _defineProperty(_messages15, 0x12, 'turn_ignition_on_off'), _messages15)
  },
  lights: {
    identifier: [0x00, 0x36],
    messages: (_messages16 = {}, _defineProperty(_messages16, 0x00, 'get_lights_state'), _defineProperty(_messages16, 0x01, 'lights_state'), _defineProperty(_messages16, 0x12, 'control_lights'), _messages16)
  },
  messaging: {
    identifier: [0x00, 0x37],
    messages: (_messages17 = {}, _defineProperty(_messages17, 0x00, 'message_received'), _defineProperty(_messages17, 0x01, 'send_message'), _messages17)
  },
  notifications: {
    identifier: [0x00, 0x38],
    messages: (_messages18 = {}, _defineProperty(_messages18, 0x00, 'notification'), _defineProperty(_messages18, 0x02, 'clear_notification'), _defineProperty(_messages18, 0x11, 'notification_action'), _messages18)
  },
  fueling: {
    identifier: [0x00, 0x40],
    messages: (_messages19 = {}, _defineProperty(_messages19, 0x00, 'get_gas_flap_state'), _defineProperty(_messages19, 0x01, 'gas_flap_state'), _defineProperty(_messages19, 0x12, 'open_close_gas_flap'), _messages19)
  },
  driver_fatigue: {
    identifier: [0x00, 0x41],
    messages: _defineProperty({}, 0x01, 'driver_fatigue_detected')
  },
  windscreen: {
    identifier: [0x00, 0x42],
    messages: (_messages21 = {}, _defineProperty(_messages21, 0x00, 'get_windscreen_state'), _defineProperty(_messages21, 0x01, 'windscreen_state'), _defineProperty(_messages21, 0x12, 'set_windscreen_damage'), _defineProperty(_messages21, 0x13, 'set_windscreen_replacement_needed'), _defineProperty(_messages21, 0x14, 'control_wipers'), _messages21)
  },
  video_handover: {
    identifier: [0x00, 0x43],
    messages: _defineProperty({}, 0x00, 'video_handover')
  },
  text_input: {
    identifier: [0x00, 0x44],
    messages: _defineProperty({}, 0x00, 'text_input')
  },
  windows: {
    identifier: [0x00, 0x45],
    messages: (_messages24 = {}, _defineProperty(_messages24, 0x00, 'get_windows_state'), _defineProperty(_messages24, 0x01, 'windows_state'), _defineProperty(_messages24, 0x12, 'control_windows'), _messages24)
  },
  theft_alarm: {
    identifier: [0x00, 0x46],
    messages: (_messages25 = {}, _defineProperty(_messages25, 0x00, 'get_theft_alarm_state'), _defineProperty(_messages25, 0x01, 'theft_alarm_state'), _defineProperty(_messages25, 0x12, 'set_theft_alarm_state'), _messages25)
  },
  parking_ticket: {
    identifier: [0x00, 0x47],
    messages: (_messages26 = {}, _defineProperty(_messages26, 0x00, 'get_parking_ticket'), _defineProperty(_messages26, 0x01, 'parking_ticket'), _defineProperty(_messages26, 0x02, 'start_parking'), _defineProperty(_messages26, 0x03, 'end_parking'), _messages26)
  },
  keyfob_position: {
    identifier: [0x00, 0x48],
    messages: (_messages27 = {}, _defineProperty(_messages27, 0x00, 'get_keyfob_position'), _defineProperty(_messages27, 0x01, 'keyfob_position'), _messages27)
  },
  browser: {
    identifier: [0x00, 0x49],
    messages: _defineProperty({}, 0x00, 'load_url')
  },
  vehicle_time: {
    identifier: [0x00, 0x50],
    messages: (_messages29 = {}, _defineProperty(_messages29, 0x00, 'get_vehicle_time'), _defineProperty(_messages29, 0x01, 'vehicle_time'), _messages29)
  },
  graphics: {
    identifier: [0x00, 0x51],
    messages: _defineProperty({}, 0x00, 'display_image')
  },
  offroad: {
    identifier: [0x00, 0x52],
    messages: (_messages31 = {}, _defineProperty(_messages31, 0x00, 'get_offroad_state'), _defineProperty(_messages31, 0x01, 'offroad_state'), _messages31)
  },
  chassis_settings: {
    identifier: [0x00, 0x53],
    messages: (_messages32 = {}, _defineProperty(_messages32, 0x00, 'get_chassis_settings'), _defineProperty(_messages32, 0x01, 'chassis_settings'), _defineProperty(_messages32, 0x12, 'set_driving_mode'), _defineProperty(_messages32, 0x13, 'start_stop_sports_chrono'), _defineProperty(_messages32, 0x14, 'set_spring_rates'), _defineProperty(_messages32, 0x15, 'set_chassis_position'), _messages32)
  },
  light_conditions: {
    identifier: [0x00, 0x54],
    messages: (_messages33 = {}, _defineProperty(_messages33, 0x00, 'get_light_conditions'), _defineProperty(_messages33, 0x01, 'light_conditions'), _messages33)
  },
  weather_conditions: {
    identifier: [0x00, 0x55],
    messages: (_messages34 = {}, _defineProperty(_messages34, 0x00, 'get_weather_conditions'), _defineProperty(_messages34, 0x01, 'weather_conditions'), _messages34)
  },
  seats: {
    identifier: [0x00, 0x56],
    messages: (_messages35 = {}, _defineProperty(_messages35, 0x00, 'get_seats_state'), _defineProperty(_messages35, 0x01, 'seats_state'), _messages35)
  },
  race: {
    identifier: [0x00, 0x57],
    messages: (_messages36 = {}, _defineProperty(_messages36, 0x00, 'get_race_state'), _defineProperty(_messages36, 0x01, 'race_state'), _messages36)
  },
  parking_brake: {
    identifier: [0x00, 0x58],
    messages: (_messages37 = {}, _defineProperty(_messages37, 0x00, 'get_parking_brake_state'), _defineProperty(_messages37, 0x01, 'parking_brake_state'), _defineProperty(_messages37, 0x12, 'set_parking_brake'), _messages37)
  },
  wi_fi: {
    identifier: [0x00, 0x59],
    messages: (_messages38 = {}, _defineProperty(_messages38, 0x00, 'get_wi_fi_state'), _defineProperty(_messages38, 0x01, 'wi_fi_state'), _defineProperty(_messages38, 0x02, 'connect_to_network'), _defineProperty(_messages38, 0x03, 'forget_network'), _defineProperty(_messages38, 0x04, 'enable_disable_wi_fi'), _messages38)
  },
  home_charger: {
    identifier: [0x00, 0x60],
    messages: (_messages39 = {}, _defineProperty(_messages39, 0x00, 'get_home_charger_state'), _defineProperty(_messages39, 0x01, 'home_charger_state'), _defineProperty(_messages39, 0x12, 'set_charge_current'), _defineProperty(_messages39, 0x13, 'set_price_tariffs'), _defineProperty(_messages39, 0x14, 'activate_deactivate_solar_charging'), _defineProperty(_messages39, 0x15, 'enable_disable_wi_fi_hotspot'), _defineProperty(_messages39, 0x16, 'authenticate_expire'), _messages39)
  },
  dashboard_lights: {
    identifier: [0x00, 0x61],
    messages: (_messages40 = {}, _defineProperty(_messages40, 0x00, 'get_dashboard_lights'), _defineProperty(_messages40, 0x01, 'dashboard_lights'), _messages40)
  },
  cruise_control: {
    identifier: [0x00, 0x62],
    messages: (_messages41 = {}, _defineProperty(_messages41, 0x00, 'get_cruise_control_state'), _defineProperty(_messages41, 0x01, 'cruise_control_state'), _defineProperty(_messages41, 0x12, 'activate_deactivate_cruise_control'), _messages41)
  },
  start_stop: {
    identifier: [0x00, 0x63],
    messages: (_messages42 = {}, _defineProperty(_messages42, 0x00, 'get_start_stop_state'), _defineProperty(_messages42, 0x01, 'start_stop_state'), _defineProperty(_messages42, 0x12, 'activate_deactivate_start_stop'), _messages42)
  },
  tachograph: {
    identifier: [0x00, 0x64],
    messages: (_messages43 = {}, _defineProperty(_messages43, 0x00, 'get_tachograph_state'), _defineProperty(_messages43, 0x01, 'tachograph_state'), _messages43)
  },
  power_takeoff: {
    identifier: [0x00, 0x65],
    messages: (_messages44 = {}, _defineProperty(_messages44, 0x00, 'get_power_takeoff_state'), _defineProperty(_messages44, 0x01, 'power_takeoff_state'), _defineProperty(_messages44, 0x02, 'activate_deactivate_power_takeoff'), _messages44)
  },
  mobile: {
    identifier: [0x00, 0x66],
    messages: (_messages45 = {}, _defineProperty(_messages45, 0x00, 'get_mobile_state'), _defineProperty(_messages45, 0x01, 'mobile_state'), _messages45)
  },
  hood: {
    identifier: [0x00, 0x67],
    messages: (_messages46 = {}, _defineProperty(_messages46, 0x00, 'get_hood_state'), _defineProperty(_messages46, 0x01, 'hood_state'), _messages46)
  },
  usage: {
    identifier: [0x00, 0x68],
    messages: (_messages47 = {}, _defineProperty(_messages47, 0x00, 'get_usage'), _defineProperty(_messages47, 0x01, 'usage'), _messages47)
  }
};
