// hmkit/es/Commands.js
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

import BrowserCommand from './Commands/BrowserCommand';
import CapabilitiesCommand from './Commands/CapabilitiesCommand';
import ChargingCommand from './Commands/ChargingCommand';
import ChassisSettingsCommand from './Commands/ChassisSettingsCommand';
import ClimateCommand from './Commands/ClimateCommand';
import CruiseControlCommand from './Commands/CruiseControlCommand';
import DashboardLightsCommand from './Commands/DashboardLightsCommand';
import DiagnosticsCommand from './Commands/DiagnosticsCommand';
import DoorLocksCommand from './Commands/DoorLocksCommand';
import EngineCommand from './Commands/EngineCommand';
import FirmwareVersionCommand from './Commands/FirmwareVersionCommand';
import FuelingCommand from './Commands/FuelingCommand';
import GraphicsCommand from './Commands/GraphicsCommand';
import HistoricalCommand from './Commands/HistoricalCommand';
import HomeChargerCommand from './Commands/HomeChargerCommand';
import HonkHornFlashLightsCommand from './Commands/HonkHornFlashLightsCommand';
import HoodCommand from './Commands/HoodCommand';
import LightConditionsCommand from './Commands/LightConditionsCommand';
import LightsCommand from './Commands/LightsCommand';
import MaintenanceCommand from './Commands/MaintenanceCommand';
import MessagingCommand from './Commands/MessagingCommand';
import MobileCommand from './Commands/MobileCommand';
import MultiCommand from './Commands/MultiCommand';
import NaviDestinationCommand from './Commands/NaviDestinationCommand';
import NotificationCommand from './Commands/NotificationCommand';
import OffroadCommand from './Commands/OffroadCommand';
import ParkingBrakeCommand from './Commands/ParkingBrakeCommand';
import ParkingTicketCommand from './Commands/ParkingTicketCommand';
import PowerTakeOffCommand from './Commands/PowerTakeOffCommand';
import RaceCommand from './Commands/RaceCommand';
import RooftopControlCommand from './Commands/RooftopControlCommand';
import SeatsCommand from './Commands/SeatsCommand';
import StartStopCommand from './Commands/StartStopCommand';
import TachographCommand from './Commands/TachographCommand';
import TextInputCommand from './Commands/TextInputCommand';
import TheftAlarmCommand from './Commands/TheftAlarmCommand';
import TrunkAccessCommand from './Commands/TrunkAccessCommand';
import UsageCommand from './Commands/UsageCommand';
import ValetModeCommand from './Commands/ValetModeCommand';
import VehicleLocationCommand from './Commands/VehicleLocationCommand';
import VehicleStatusCommand from './Commands/VehicleStatusCommand';
import VehicleTimeCommand from './Commands/VehicleTimeCommand';
import VideoHandoverCommand from './Commands/VideoHandoverCommand';
import WakeUpCommand from './Commands/WakeUpCommand';
import WeatherConditionsCommand from './Commands/WeatherConditionsCommand';
import WiFiCommand from './Commands/WiFiCommand';
import WindowsCommand from './Commands/WindowsCommand';
import WindscreenCommand from './Commands/WindscreenCommand';

var Commands = function Commands() {
  _classCallCheck(this, Commands);

  Object.assign(this, {
    BrowserCommand: BrowserCommand,
    CapabilitiesCommand: CapabilitiesCommand,
    ChargingCommand: ChargingCommand,
    ChassisSettingsCommand: ChassisSettingsCommand,
    ClimateCommand: ClimateCommand,
    CruiseControlCommand: CruiseControlCommand,
    DashboardLightsCommand: DashboardLightsCommand,
    DiagnosticsCommand: DiagnosticsCommand,
    DoorLocksCommand: DoorLocksCommand,
    EngineCommand: EngineCommand,
    FirmwareVersionCommand: FirmwareVersionCommand,
    FuelingCommand: FuelingCommand,
    GraphicsCommand: GraphicsCommand,
    HistoricalCommand: HistoricalCommand,
    HomeChargerCommand: HomeChargerCommand,
    HonkHornFlashLightsCommand: HonkHornFlashLightsCommand,
    HoodCommand: HoodCommand,
    LightConditionsCommand: LightConditionsCommand,
    LightsCommand: LightsCommand,
    MaintenanceCommand: MaintenanceCommand,
    MessagingCommand: MessagingCommand,
    MobileCommand: MobileCommand,
    MultiCommand: MultiCommand,
    NaviDestinationCommand: NaviDestinationCommand,
    NotificationCommand: NotificationCommand,
    OffroadCommand: OffroadCommand,
    ParkingBrakeCommand: ParkingBrakeCommand,
    ParkingTicketCommand: ParkingTicketCommand,
    PowerTakeOffCommand: PowerTakeOffCommand,
    RaceCommand: RaceCommand,
    RooftopControlCommand: RooftopControlCommand,
    SeatsCommand: SeatsCommand,
    StartStopCommand: StartStopCommand,
    TachographCommand: TachographCommand,
    TextInputCommand: TextInputCommand,
    TheftAlarmCommand: TheftAlarmCommand,
    TrunkAccessCommand: TrunkAccessCommand,
    UsageCommand: UsageCommand,
    ValetModeCommand: ValetModeCommand,
    VehicleLocationCommand: VehicleLocationCommand,
    VehicleStatusCommand: VehicleStatusCommand,
    VehicleTimeCommand: VehicleTimeCommand,
    VideoHandoverCommand: VideoHandoverCommand,
    WakeUpCommand: WakeUpCommand,
    WeatherConditionsCommand: WeatherConditionsCommand,
    WiFiCommand: WiFiCommand,
    WindowsCommand: WindowsCommand,
    WindscreenCommand: WindscreenCommand
  });
};

export default Commands;
