// hmkit/es/PropertyResponse.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

import Property from './Property';
import { bytesSum } from './helpers';

var PropertyResponse = function () {
  function PropertyResponse() {
    _classCallCheck(this, PropertyResponse);
  }

  _createClass(PropertyResponse, [{
    key: 'parse',

    /*
     * parse()
     *
     * data - incoming response data
     * properties - capability properties that should be extracted from response
     *
     * This parses all configured properties and adds them to "this" by their namespace.
     * Properties that do not have value will be ignored.
     */
    value: function parse(data, properties) {
      this.bindProperties(this.parseProperties(data, properties).filter(function (property) {
        return property.value !== null && property.value !== undefined || property.subProperties.length > 0;
      }));
    }

    /*
     * bindProperties()
     *
     * properties - parsed properties with values that will be binded.
     *
     * Binds properties to "this" by their namespace.
     * This function can be used in extending class to override binding.
     */

  }, {
    key: 'bindProperties',
    value: function bindProperties(properties) {
      var _this = this;

      properties.forEach(function (property) {
        _this[property.namespace] = property.getValue();
      });
    }

    /*
     * parseProperties()
     *
     * data - incoming response data
     * properties - capability properties that should be extracted from response
     *
     * Parses data and sets decoded values for each property that is specified in response.
     * Returns properties with values. If property data was not in the response data then property will not have a value.
     *
     * TODO: Add data length validation and individual property length validation
     * TODO: Maybe split this function into more readable chunks. This function does too much at the moment.
     */

  }, {
    key: 'parseProperties',
    value: function parseProperties(data, properties) {
      var propertiesData = [].concat(_toConsumableArray(data.slice(3, data.length)));

      if (propertiesData.length > 0) {
        var counter = 0;

        while (counter < propertiesData.length) {
          var identifier = propertiesData[counter];
          var propertyDataLength = bytesSum(propertiesData.slice(counter + 1, counter + 3));
          var propertyData = propertiesData.slice(counter + 3, counter + 3 + propertyDataLength);

          var property = this.findProperty(identifier, properties);

          if (!!property) {
            property.parseValue(propertyData);
          }

          counter += 3 + propertyDataLength;
        }
      }

      return properties;
    }

    /*
     * findProperty()
     *
     * identifier - property identifier
     * properties - list of properties where to search from
     *
     * Simple array find helper to keep the code cleaner.
     * Returns property with correct identifier or null if property was not found.
     * If property is not found, parent function should definitely throw/show error.
     */

  }, {
    key: 'findProperty',
    value: function findProperty(identifier, properties) {
      return properties.find(function (property) {
        return property.identifier === identifier;
      });
    }
  }]);

  return PropertyResponse;
}();

export default PropertyResponse;
