// hmkit/lib/Commands.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _BrowserCommand = require('./Commands/BrowserCommand');

var _BrowserCommand2 = _interopRequireDefault(_BrowserCommand);

var _CapabilitiesCommand = require('./Commands/CapabilitiesCommand');

var _CapabilitiesCommand2 = _interopRequireDefault(_CapabilitiesCommand);

var _ChargingCommand = require('./Commands/ChargingCommand');

var _ChargingCommand2 = _interopRequireDefault(_ChargingCommand);

var _ChassisSettingsCommand = require('./Commands/ChassisSettingsCommand');

var _ChassisSettingsCommand2 = _interopRequireDefault(_ChassisSettingsCommand);

var _ClimateCommand = require('./Commands/ClimateCommand');

var _ClimateCommand2 = _interopRequireDefault(_ClimateCommand);

var _CruiseControlCommand = require('./Commands/CruiseControlCommand');

var _CruiseControlCommand2 = _interopRequireDefault(_CruiseControlCommand);

var _DashboardLightsCommand = require('./Commands/DashboardLightsCommand');

var _DashboardLightsCommand2 = _interopRequireDefault(_DashboardLightsCommand);

var _DiagnosticsCommand = require('./Commands/DiagnosticsCommand');

var _DiagnosticsCommand2 = _interopRequireDefault(_DiagnosticsCommand);

var _DoorLocksCommand = require('./Commands/DoorLocksCommand');

var _DoorLocksCommand2 = _interopRequireDefault(_DoorLocksCommand);

var _EngineCommand = require('./Commands/EngineCommand');

var _EngineCommand2 = _interopRequireDefault(_EngineCommand);

var _FirmwareVersionCommand = require('./Commands/FirmwareVersionCommand');

var _FirmwareVersionCommand2 = _interopRequireDefault(_FirmwareVersionCommand);

var _FuelingCommand = require('./Commands/FuelingCommand');

var _FuelingCommand2 = _interopRequireDefault(_FuelingCommand);

var _GraphicsCommand = require('./Commands/GraphicsCommand');

var _GraphicsCommand2 = _interopRequireDefault(_GraphicsCommand);

var _HistoricalCommand = require('./Commands/HistoricalCommand');

var _HistoricalCommand2 = _interopRequireDefault(_HistoricalCommand);

var _HomeChargerCommand = require('./Commands/HomeChargerCommand');

var _HomeChargerCommand2 = _interopRequireDefault(_HomeChargerCommand);

var _HonkHornFlashLightsCommand = require('./Commands/HonkHornFlashLightsCommand');

var _HonkHornFlashLightsCommand2 = _interopRequireDefault(_HonkHornFlashLightsCommand);

var _HoodCommand = require('./Commands/HoodCommand');

var _HoodCommand2 = _interopRequireDefault(_HoodCommand);

var _LightConditionsCommand = require('./Commands/LightConditionsCommand');

var _LightConditionsCommand2 = _interopRequireDefault(_LightConditionsCommand);

var _LightsCommand = require('./Commands/LightsCommand');

var _LightsCommand2 = _interopRequireDefault(_LightsCommand);

var _MaintenanceCommand = require('./Commands/MaintenanceCommand');

var _MaintenanceCommand2 = _interopRequireDefault(_MaintenanceCommand);

var _MessagingCommand = require('./Commands/MessagingCommand');

var _MessagingCommand2 = _interopRequireDefault(_MessagingCommand);

var _MobileCommand = require('./Commands/MobileCommand');

var _MobileCommand2 = _interopRequireDefault(_MobileCommand);

var _MultiCommand = require('./Commands/MultiCommand');

var _MultiCommand2 = _interopRequireDefault(_MultiCommand);

var _NaviDestinationCommand = require('./Commands/NaviDestinationCommand');

var _NaviDestinationCommand2 = _interopRequireDefault(_NaviDestinationCommand);

var _NotificationCommand = require('./Commands/NotificationCommand');

var _NotificationCommand2 = _interopRequireDefault(_NotificationCommand);

var _OffroadCommand = require('./Commands/OffroadCommand');

var _OffroadCommand2 = _interopRequireDefault(_OffroadCommand);

var _ParkingBrakeCommand = require('./Commands/ParkingBrakeCommand');

var _ParkingBrakeCommand2 = _interopRequireDefault(_ParkingBrakeCommand);

var _ParkingTicketCommand = require('./Commands/ParkingTicketCommand');

var _ParkingTicketCommand2 = _interopRequireDefault(_ParkingTicketCommand);

var _PowerTakeOffCommand = require('./Commands/PowerTakeOffCommand');

var _PowerTakeOffCommand2 = _interopRequireDefault(_PowerTakeOffCommand);

var _RaceCommand = require('./Commands/RaceCommand');

var _RaceCommand2 = _interopRequireDefault(_RaceCommand);

var _RooftopControlCommand = require('./Commands/RooftopControlCommand');

var _RooftopControlCommand2 = _interopRequireDefault(_RooftopControlCommand);

var _SeatsCommand = require('./Commands/SeatsCommand');

var _SeatsCommand2 = _interopRequireDefault(_SeatsCommand);

var _StartStopCommand = require('./Commands/StartStopCommand');

var _StartStopCommand2 = _interopRequireDefault(_StartStopCommand);

var _TachographCommand = require('./Commands/TachographCommand');

var _TachographCommand2 = _interopRequireDefault(_TachographCommand);

var _TextInputCommand = require('./Commands/TextInputCommand');

var _TextInputCommand2 = _interopRequireDefault(_TextInputCommand);

var _TheftAlarmCommand = require('./Commands/TheftAlarmCommand');

var _TheftAlarmCommand2 = _interopRequireDefault(_TheftAlarmCommand);

var _TrunkAccessCommand = require('./Commands/TrunkAccessCommand');

var _TrunkAccessCommand2 = _interopRequireDefault(_TrunkAccessCommand);

var _UsageCommand = require('./Commands/UsageCommand');

var _UsageCommand2 = _interopRequireDefault(_UsageCommand);

var _ValetModeCommand = require('./Commands/ValetModeCommand');

var _ValetModeCommand2 = _interopRequireDefault(_ValetModeCommand);

var _VehicleLocationCommand = require('./Commands/VehicleLocationCommand');

var _VehicleLocationCommand2 = _interopRequireDefault(_VehicleLocationCommand);

var _VehicleStatusCommand = require('./Commands/VehicleStatusCommand');

var _VehicleStatusCommand2 = _interopRequireDefault(_VehicleStatusCommand);

var _VehicleTimeCommand = require('./Commands/VehicleTimeCommand');

var _VehicleTimeCommand2 = _interopRequireDefault(_VehicleTimeCommand);

var _VideoHandoverCommand = require('./Commands/VideoHandoverCommand');

var _VideoHandoverCommand2 = _interopRequireDefault(_VideoHandoverCommand);

var _WakeUpCommand = require('./Commands/WakeUpCommand');

var _WakeUpCommand2 = _interopRequireDefault(_WakeUpCommand);

var _WeatherConditionsCommand = require('./Commands/WeatherConditionsCommand');

var _WeatherConditionsCommand2 = _interopRequireDefault(_WeatherConditionsCommand);

var _WiFiCommand = require('./Commands/WiFiCommand');

var _WiFiCommand2 = _interopRequireDefault(_WiFiCommand);

var _WindowsCommand = require('./Commands/WindowsCommand');

var _WindowsCommand2 = _interopRequireDefault(_WindowsCommand);

var _WindscreenCommand = require('./Commands/WindscreenCommand');

var _WindscreenCommand2 = _interopRequireDefault(_WindscreenCommand);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Commands = function Commands() {
  _classCallCheck(this, Commands);

  Object.assign(this, {
    BrowserCommand: _BrowserCommand2.default,
    CapabilitiesCommand: _CapabilitiesCommand2.default,
    ChargingCommand: _ChargingCommand2.default,
    ChassisSettingsCommand: _ChassisSettingsCommand2.default,
    ClimateCommand: _ClimateCommand2.default,
    CruiseControlCommand: _CruiseControlCommand2.default,
    DashboardLightsCommand: _DashboardLightsCommand2.default,
    DiagnosticsCommand: _DiagnosticsCommand2.default,
    DoorLocksCommand: _DoorLocksCommand2.default,
    EngineCommand: _EngineCommand2.default,
    FirmwareVersionCommand: _FirmwareVersionCommand2.default,
    FuelingCommand: _FuelingCommand2.default,
    GraphicsCommand: _GraphicsCommand2.default,
    HistoricalCommand: _HistoricalCommand2.default,
    HomeChargerCommand: _HomeChargerCommand2.default,
    HonkHornFlashLightsCommand: _HonkHornFlashLightsCommand2.default,
    HoodCommand: _HoodCommand2.default,
    LightConditionsCommand: _LightConditionsCommand2.default,
    LightsCommand: _LightsCommand2.default,
    MaintenanceCommand: _MaintenanceCommand2.default,
    MessagingCommand: _MessagingCommand2.default,
    MobileCommand: _MobileCommand2.default,
    MultiCommand: _MultiCommand2.default,
    NaviDestinationCommand: _NaviDestinationCommand2.default,
    NotificationCommand: _NotificationCommand2.default,
    OffroadCommand: _OffroadCommand2.default,
    ParkingBrakeCommand: _ParkingBrakeCommand2.default,
    ParkingTicketCommand: _ParkingTicketCommand2.default,
    PowerTakeOffCommand: _PowerTakeOffCommand2.default,
    RaceCommand: _RaceCommand2.default,
    RooftopControlCommand: _RooftopControlCommand2.default,
    SeatsCommand: _SeatsCommand2.default,
    StartStopCommand: _StartStopCommand2.default,
    TachographCommand: _TachographCommand2.default,
    TextInputCommand: _TextInputCommand2.default,
    TheftAlarmCommand: _TheftAlarmCommand2.default,
    TrunkAccessCommand: _TrunkAccessCommand2.default,
    UsageCommand: _UsageCommand2.default,
    ValetModeCommand: _ValetModeCommand2.default,
    VehicleLocationCommand: _VehicleLocationCommand2.default,
    VehicleStatusCommand: _VehicleStatusCommand2.default,
    VehicleTimeCommand: _VehicleTimeCommand2.default,
    VideoHandoverCommand: _VideoHandoverCommand2.default,
    WakeUpCommand: _WakeUpCommand2.default,
    WeatherConditionsCommand: _WeatherConditionsCommand2.default,
    WiFiCommand: _WiFiCommand2.default,
    WindowsCommand: _WindowsCommand2.default,
    WindscreenCommand: _WindscreenCommand2.default
  });
};

exports.default = Commands;
