// hmkit/lib/encoding.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PROPERTY_FAILURE_ID = exports.PROPERTY_TIMESTAMP_ID = exports.PROPERTY_DATA_ID = undefined;
exports.percentageToDouble = percentageToDouble;
exports.base64ToUint8 = base64ToUint8;
exports.byteArrayToBase64 = byteArrayToBase64;
exports.asciiToUint8 = asciiToUint8;
exports.uint8ArrayToHex = uint8ArrayToHex;
exports.hexToInt = hexToInt;
exports.intToHex = intToHex;
exports.intToBinary = intToBinary;
exports.uint8toInt8 = uint8toInt8;
exports.binaryToInt = binaryToInt;
exports.hexToUint8Array = hexToUint8Array;
exports.hexToByteArrays = hexToByteArrays;
exports.pad = pad;
exports.hexArrayToHex = hexArrayToHex;
exports.base10ToIeee754 = base10ToIeee754;
exports.base10ToIeee754Double = base10ToIeee754Double;
exports.ieee754ToBase10 = ieee754ToBase10;
exports.ieee754DoubleToBase10 = ieee754DoubleToBase10;
exports.stringToHex = stringToHex;
exports.stringToBytes = stringToBytes;
exports.intToTwoBytes = intToTwoBytes;
exports.dateToBytes = dateToBytes;
exports.bytesToString = bytesToString;
exports.decimalToHexArray = decimalToHexArray;
exports.utfStringToByteArray = utfStringToByteArray;

var _atob = require('atob');

var _atob2 = _interopRequireDefault(_atob);

var _btoa = require('btoa');

var _btoa2 = _interopRequireDefault(_btoa);

var _ieee = require('ieee754');

var _ieee2 = _interopRequireDefault(_ieee);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PROPERTY_DATA_ID = exports.PROPERTY_DATA_ID = 0x01;
var PROPERTY_TIMESTAMP_ID = exports.PROPERTY_TIMESTAMP_ID = 0x02;
var PROPERTY_FAILURE_ID = exports.PROPERTY_FAILURE_ID = 0x03;

function percentageToDouble(value) {
  return base10ToIeee754Double(value);
}

function base64ToUint8(base64String) {
  return new Uint8Array((0, _atob2.default)(base64String).split('').map(function (c) {
    return c.charCodeAt(0);
  }));
}

function byteArrayToBase64(byteArray) {
  var bytes = new Uint8Array(byteArray);
  return (0, _btoa2.default)(String.fromCharCode.apply(null, bytes));
}

function asciiToUint8(string) {
  return new Uint8Array(string.split('').map(function (c) {
    return c.charCodeAt(0);
  }));
}

function uint8ArrayToHex(uint8Array) {
  return uint8Array.reduce(function (memo, i) {
    return memo + pad(i.toString(16), 2);
  }, '');
}

function hexToInt(hex) {
  return parseInt(hex, 16);
}

function intToHex(int) {
  return int.toString(16);
}

function intToBinary(int) {
  return int.toString(2);
}

function uint8toInt8(uint8) {
  return uint8 << 24 >> 24;
}

function binaryToInt(binary) {
  return parseInt(binary, 2);
}

function hexToUint8Array(hexString) {
  if (!hexString) {
    return new Uint8Array();
  }

  var byteArray = [];

  for (var i = 0, len = hexString.length; i < len; i += 2) {
    byteArray.push(parseInt(hexString.substr(i, 2), 16));
  }

  return new Uint8Array(byteArray);
}

function hexToByteArrays(hexString) {
  var uint8Array = hexToUint8Array(hexString);
  var byteArrays = [];

  uint8Array.forEach(function (uint8) {
    byteArrays.push(pad(uint8.toString(2), 8).split('').map(function (byte) {
      return Number(byte);
    }));
  });

  return byteArrays;
}

function pad(string, width) {
  return string.length >= width ? string : new Array(width - string.length + 1).join('0') + string;
}

function hexArrayToHex(hexArray) {
  return hexArray.reduce(function (memo, i) {
    return memo + pad(i.toString(16), 2);
  }, '');
}

function base10ToIeee754(value) {
  var bytes = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 4;

  var ieeeArray = [];
  _ieee2.default.write(ieeeArray, value, 0, false, 23, bytes);
  return ieeeArray;
}

function base10ToIeee754Double(value) {
  var bytes = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 8;

  var ieeeArray = [];
  _ieee2.default.write(ieeeArray, value, 0, false, 52, bytes);
  return ieeeArray;
}

function ieee754ToBase10(array) {
  var bytes = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 4;

  return _ieee2.default.read(array, 0, false, 23, bytes);
}

function ieee754DoubleToBase10(array) {
  var bytes = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 8;

  return _ieee2.default.read(array, 0, false, 52, bytes);
}

function stringToHex(string) {
  var result = '';
  for (var i = 0; i < string.length; i++) {
    result += string[i].charCodeAt(0).toString(16);
  }
  return result;
}

function stringToBytes(string) {
  return hexToUint8Array(stringToHex(string));
}

function intToTwoBytes(int) {
  return hexToUint8Array(pad(intToHex(int), 4));
}

function dateToBytes(date) {
  return decimalToHexArray(date.getTime());
}

// This also removes null bytes
function bytesToString(bytes) {
  return Buffer.from(bytes).toString('utf8').replace(/\0/g, '');
}

function decimalToHexArray(value) {
  var bytes = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 1;

  var hex = parseInt(value, 10).toString(16);
  while (hex.length % (bytes * 2) !== 0) {
    hex = '0' + hex;
  }var hexArray = hex.match(/.{1,2}/g);

  return hexArray.map(function (hexItem) {
    return Number('0x' + hexItem);
  });
}

function utfStringToByteArray(string, minLength) {
  var byteArray = unescape(encodeURIComponent(string)).split('').map(function (char) {
    return char.charCodeAt(0);
  });

  if (minLength !== undefined) {
    while (byteArray.length < minLength) {
      byteArray.unshift(0);
    }
  }

  return byteArray;
}
