// hmkit/lib/Property.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Property = function Property(identifier, namespace) {
  _classCallCheck(this, Property);

  _initialiseProps.call(this);

  this.identifier = identifier;
  this.namespace = namespace;
  this.value = undefined;
  this.isArray = false;
  this.subProperties = [];
  this.subPropertiesIdentifierNamespace = null;
};

var _initialiseProps = function _initialiseProps() {
  var _this = this;

  this.array = function () {
    _this.isArray = true;
    return _this;
  };

  this.getValue = function () {
    if (_this.subProperties.length > 0 && _this.subPropertiesIdentifierNamespace !== null) {
      var value = [];

      _this.subProperties.forEach(function (subProperty) {
        var subPropertyValue = subProperty.getValue();

        if (subPropertyValue !== undefined && subPropertyValue !== null) {
          value.push(_extends(_defineProperty({}, _this.subPropertiesIdentifierNamespace, subProperty.identifierValue), subPropertyValue));
        }
      });

      return value;
    }

    return _this.value;
  };

  this.parseValue = function (data) {
    if (_this.subProperties.length > 0 && data.length > 0) {
      var subProperty = _this.findSubProperty(data[0]);

      if (!!subProperty) {
        subProperty.parseValue(data.slice(1, data.length));
      }

      return null;
    }

    if (_this.isArray) {
      if (!Array.isArray(_this.value)) _this.value = [];

      _this.value.push(_this.decode(data));
    } else {
      _this.value = _this.decode(data);
    }

    return _this.value;
  };

  this.decode = function (data) {
    if (!!_this.decoder) {
      return _this.decoder(data);
    } else if (data.length === 1) {
      return data[0];
    }

    return data;
  };

  this.setDecoder = function (decoder) {
    _this.decoder = decoder;
    return _this;
  };

  this.setOptionalSubProperties = function (identifierNamespace, subProperties) {
    var _subProperties;

    (_subProperties = _this.subProperties).push.apply(_subProperties, _toConsumableArray(subProperties));
    _this.subPropertiesIdentifierNamespace = identifierNamespace;
    return _this;
  };

  this.findSubProperty = function (identifier) {
    return _this.subProperties.find(function (supProperty) {
      return supProperty.identifier === identifier;
    });
  };
};

exports.default = Property;
