// hmkit/lib/Responses/DoorLocksResponse.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _PropertyResponse2 = require('../PropertyResponse');

var _PropertyResponse3 = _interopRequireDefault(_PropertyResponse2);

var _Property = require('../Property');

var _Property2 = _interopRequireDefault(_Property);

var _OptionalProperty = require('../OptionalProperty');

var _OptionalProperty2 = _interopRequireDefault(_OptionalProperty);

var _helpers = require('../helpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var DoorLocksResponse = function (_PropertyResponse) {
  _inherits(DoorLocksResponse, _PropertyResponse);

  /**
   * @property {Array} insideLocks (array) Inside door locks ([{ doorLocation: (string 'front_left|front_right|rear_right|rear_left|hatch|all'), lockState: (string 'unlocked|locked') }])
   * @property {Array} locks (array) Door locks ([{ doorLocation: (string 'front_left|front_right|rear_right|rear_left|hatch|all'), lockState: (string 'unlocked|locked') }])
   * @property {Array} positions (Array { doorLocation: (string 'front_left|front_right|rear_right|rear_left|hatch|all'), position: (string 'open|closed')}`) Positions
   *
   * @example DoorLocksResponse
    {
      insideLocks: [{
        doorLocation: 'front_left',
        lockState: 'unlocked'
      }, {
        doorLocation: 'front_right',
        lockState: 'unlocked'
      }, {
        doorLocation: 'rear_right',
        lockState: 'unlocked'
      }, {
        doorLocation: 'rear_left',
        lockState: 'unlocked'
      }, {
        doorLocation: 'hatch',
        lockState: 'unlocked'
      }, {
        doorLocation: 'all',
        lockState: 'unlocked'
      }],
      locks: [{
        doorLocation: 'front_left',
        lockState: 'unlocked'
      }, {
        doorLocation: 'front_right',
        lockState: 'unlocked'
      }, {
        doorLocation: 'rear_right',
        lockState: 'unlocked'
      }, {
        doorLocation: 'rear_left',
        lockState: 'unlocked'
      }, {
        doorLocation: 'hatch',
        lockState: 'unlocked'
      }, {
        doorLocation: 'all',
        lockState: 'unlocked'
      }],
      positions: [{
        doorLocation: 'front_left',
        position: 'closed'
      }, {
        doorLocation: 'front_right',
        position: 'closed'
      }, {
        doorLocation: 'rear_right',
        position: 'open'
      }, {
        doorLocation: 'rear_left',
        position: 'closed'
      }, {
        doorLocation: 'hatch',
        position: 'closed'
      }, {
        doorLocation: 'all',
        position: 'closed'
      }]
    }
    */
  function DoorLocksResponse(data) {
    _classCallCheck(this, DoorLocksResponse);

    /* prettier-ignore */
    var _this = _possibleConstructorReturn(this, (DoorLocksResponse.__proto__ || Object.getPrototypeOf(DoorLocksResponse)).call(this));

    var properties = [new _Property2.default(0x02, 'insideLocks').setOptionalSubProperties('doorLocation', [new _OptionalProperty2.default(0x00, 'front_left').setDecoder(_this.lockDecoder), new _OptionalProperty2.default(0x01, 'front_right').setDecoder(_this.lockDecoder), new _OptionalProperty2.default(0x02, 'rear_right').setDecoder(_this.lockDecoder), new _OptionalProperty2.default(0x03, 'rear_left').setDecoder(_this.lockDecoder), new _OptionalProperty2.default(0x04, 'hatch').setDecoder(_this.lockDecoder), new _OptionalProperty2.default(0x05, 'all').setDecoder(_this.lockDecoder)]), new _Property2.default(0x03, 'locks').setOptionalSubProperties('doorLocation', [new _OptionalProperty2.default(0x00, 'front_left').setDecoder(_this.lockDecoder), new _OptionalProperty2.default(0x01, 'front_right').setDecoder(_this.lockDecoder), new _OptionalProperty2.default(0x02, 'rear_right').setDecoder(_this.lockDecoder), new _OptionalProperty2.default(0x03, 'rear_left').setDecoder(_this.lockDecoder), new _OptionalProperty2.default(0x04, 'hatch').setDecoder(_this.lockDecoder), new _OptionalProperty2.default(0x05, 'all').setDecoder(_this.lockDecoder)]), new _Property2.default(0x04, 'positions').setOptionalSubProperties('doorLocation', [new _OptionalProperty2.default(0x00, 'front_left').setDecoder(_this.positionDecoder), new _OptionalProperty2.default(0x01, 'front_right').setDecoder(_this.positionDecoder), new _OptionalProperty2.default(0x02, 'rear_right').setDecoder(_this.positionDecoder), new _OptionalProperty2.default(0x03, 'rear_left').setDecoder(_this.positionDecoder), new _OptionalProperty2.default(0x04, 'hatch').setDecoder(_this.positionDecoder), new _OptionalProperty2.default(0x05, 'all').setDecoder(_this.positionDecoder)])];
    /* prettier-ignore-end */

    _this.parse(data, properties);
    return _this;
  }

  _createClass(DoorLocksResponse, [{
    key: 'positionDecoder',
    value: function positionDecoder(data) {
      return {
        position: (0, _helpers.switchDecoder)({
          0x00: 'closed',
          0x01: 'open'
        })(data)
      };
    }
  }, {
    key: 'lockDecoder',
    value: function lockDecoder(data) {
      return {
        lockState: (0, _helpers.switchDecoder)({
          0x00: 'unlocked',
          0x01: 'locked'
        })(data)
      };
    }
  }]);

  return DoorLocksResponse;
}(_PropertyResponse3.default);

DoorLocksResponse.identifier = [0x00, 0x20];
exports.default = DoorLocksResponse;
