// hmkit/lib/Responses/HomeChargerResponse.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _PropertyResponse2 = require('../PropertyResponse');

var _PropertyResponse3 = _interopRequireDefault(_PropertyResponse2);

var _Property = require('../Property');

var _Property2 = _interopRequireDefault(_Property);

var _OptionalProperty = require('../OptionalProperty');

var _OptionalProperty2 = _interopRequireDefault(_OptionalProperty);

var _encoding = require('../encoding');

var _helpers = require('../helpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var HomeChargerResponse = function (_PropertyResponse) {
  _inherits(HomeChargerResponse, _PropertyResponse);

  /**
   * @property {String} charging (string: 'disconnected|plugged_in|charging') Charging
   * @property {String} authenticationMechanism (string 'pin|app') Authentication mechanism
   * @property {String} plugType (string: 'type_1|type_2|ccs|chademo') Plug type
   * @property {Number} chargingPower (number) Charging power in kW formatted in 4-bytes per IEEE 754
   * @property {String} solarCharging (string: 'activated|deactivated') Solar charging
   * @property {String} hotspotEnabled (string: 'disabled|enabled') Hotspot enabled
   * @property {String} hotspotSSID (string) Wi-Fi Hotspot SSID formatted in UTF-8
   * @property {String} wiFiHotspotSecurity (string: 'none|wep|wpa|wpa2_personal') Wi-Fi hotspot security
   * @property {String} wiFiHotspotPassword (string) Wi-Fi Hotspot password formatted in UTF-8
   * @property {String} authentication (string: 'unauthenticated|authenticated') Authentication state
   * @property {Number} chargeCurrentDC (number) The charge current (DC) per IEEE 754
   * @property {Number} maximumChargeCurrent (number) The maximum possible charge current per IEEE 754
   * @property {Number} minimumChargeCurrent (number) The minimal possible charge current per IEEE 754
   * @property {Object} coordinates (object) Coordinates ({ latitude: (double), longitude: (double) })
   * @property {Object} priceTariffs (object) Price tariffs ({ pricingType: (string 'starting_fee|per_minute|per_kwh'), currency: (string), price: (number) })
   *
   * @example HomeChargerResponse
    {
      charging: 'disconnected',
      authenticationMechanism: 'pin',
      plugType: 'type_1',
      chargingPower: 0,
      solarCharging: 'deactivated',
      hotspotEnabled: 'disabled',
      hotspotSSID: 'SSID123123',
      wiFiHotspotSecurity: 'none',
      wiFiHotspotPassword: 'PASSWORD',
      authentication: 'unauthenticated',
      chargeCurrentDC: 0.6,
      maximumChargeCurrent: 50,
      minimumChargeCurrent: 0,
      coordinates: {
        latitude: 52.52,
        longitude: 13.41
      },
      priceTariffs: [{
        pricingType: 'starting_fee',
        price: 2,
        currency: 'EUR'
      }, {
        pricingType: 'per_minute',
        price: 1,
        currency: 'EUR'
      }, {
        pricingType: 'per_kwh',
        price: 3,
        currency: 'EUR'
      }]
    }
    */
  function HomeChargerResponse(data) {
    _classCallCheck(this, HomeChargerResponse);

    var _this = _possibleConstructorReturn(this, (HomeChargerResponse.__proto__ || Object.getPrototypeOf(HomeChargerResponse)).call(this));

    var properties = [new _Property2.default(0x01, 'charging').setDecoder((0, _helpers.switchDecoder)({
      0x00: 'disconnected',
      0x01: 'plugged_in',
      0x02: 'charging'
    })), new _Property2.default(0x02, 'authenticationMechanism').setDecoder((0, _helpers.switchDecoder)({
      0x00: 'pin',
      0x01: 'app'
    })), new _Property2.default(0x03, 'plugType').setDecoder((0, _helpers.switchDecoder)({
      0x00: 'type_1',
      0x01: 'type_2',
      0x02: 'ccs',
      0x03: 'chademo'
    })), new _Property2.default(0x04, 'chargingPower').setDecoder((0, _helpers.getRoundedIeee754ToBase10)(2)), new _Property2.default(0x05, 'solarCharging').setDecoder((0, _helpers.switchDecoder)({
      0x00: 'deactivated',
      0x01: 'activated'
    })), new _Property2.default(0x08, 'hotspotEnabled').setDecoder((0, _helpers.switchDecoder)({
      0x00: 'disabled',
      0x01: 'enabled'
    })), new _Property2.default(0x09, 'hotspotSSID').setDecoder(_encoding.bytesToString), new _Property2.default(0x0a, 'wiFiHotspotSecurity').setDecoder((0, _helpers.switchDecoder)({
      0x00: 'none',
      0x01: 'wep',
      0x02: 'wpa',
      0x03: 'wpa2_personal'
    })), new _Property2.default(0x0b, 'wiFiHotspotPassword').setDecoder(_encoding.bytesToString), new _Property2.default(0x0d, 'authentication').setDecoder((0, _helpers.switchDecoder)({
      0x00: 'unauthenticated',
      0x01: 'authenticated'
    })), new _Property2.default(0x0e, 'chargeCurrentDC').setDecoder((0, _helpers.getRoundedIeee754ToBase10)(2)), new _Property2.default(0x0f, 'maximumChargeCurrent').setDecoder((0, _helpers.getRoundedIeee754ToBase10)(2)), new _Property2.default(0x10, 'minimumChargeCurrent').setDecoder((0, _helpers.getRoundedIeee754ToBase10)(2)), new _Property2.default(0x11, 'coordinates').setDecoder(_helpers.coordinatesDecoder), new _Property2.default(0x12, 'priceTariffs').setOptionalSubProperties('pricingType', [new _OptionalProperty2.default(0x00, 'starting_fee').setDecoder(_this.priceTariffDecoder), new _OptionalProperty2.default(0x01, 'per_minute').setDecoder(_this.priceTariffDecoder), new _OptionalProperty2.default(0x02, 'per_kwh').setDecoder(_this.priceTariffDecoder)])];

    _this.parse(data, properties);
    return _this;
  }

  _createClass(HomeChargerResponse, [{
    key: 'priceTariffDecoder',
    value: function priceTariffDecoder(bytes) {
      var decoder = (0, _helpers.getRoundedIeee754ToBase10)(2);

      return {
        price: decoder(bytes.slice(0, 4)),
        currency: (0, _encoding.bytesToString)(bytes.slice(4, 7))
      };
    }
  }]);

  return HomeChargerResponse;
}(_PropertyResponse3.default);

HomeChargerResponse.identifier = [0x00, 0x60];
exports.default = HomeChargerResponse;
