// hmkit/lib/Responses/ParkingTicketResponse.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _PropertyResponse2 = require('../PropertyResponse');

var _PropertyResponse3 = _interopRequireDefault(_PropertyResponse2);

var _PropertyDecoder = require('../PropertyDecoder');

var _PropertyDecoder2 = _interopRequireDefault(_PropertyDecoder);

var _encoding = require('../encoding');

var _helpers = require('../helpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ParkingTicketResponse = function (_PropertyResponse) {
  _inherits(ParkingTicketResponse, _PropertyResponse);

  /**
   * @property {String} state (string) Parking Ticket state
   * @property {String} operatorName (string) Operator name bytes formatted in UTF-8
   * @property {String} operatorTicketID (string) Operator ticket id
   * @property {Date} startDate (date) Parking ticket start time
   * @property {Date} endDate (date) Parking ticket end time
   *
   * @example ParkingTicketResponse
    {
      parkingTicketState: {
        value: 'ended',
      },
      operatorName: {
        value: 'Berlin Parking',
      },
      operatorTicketID: {
        value: '64894233',
      },
      ticketStartTime: {
        value: '2018-04-12T23:20:50.000Z',
      },
      ticketEndTime: {
        value: '2019-03-04T09:25:58.360Z',
      },
    }
   */
  function ParkingTicketResponse(data, config) {
    _classCallCheck(this, ParkingTicketResponse);

    var _this = _possibleConstructorReturn(this, (ParkingTicketResponse.__proto__ || Object.getPrototypeOf(ParkingTicketResponse)).call(this));

    var properties = [new _PropertyDecoder2.default(0x01, 'parkingTicketState').setDecoder((0, _helpers.switchDecoder)({
      0x00: 'ended',
      0x01: 'started'
    })), new _PropertyDecoder2.default(0x02, 'operatorName').setDecoder(_encoding.bytesToString), new _PropertyDecoder2.default(0x03, 'operatorTicketID').setDecoder(_encoding.bytesToString), new _PropertyDecoder2.default(0x04, 'ticketStartTime').setDecoder(_helpers.timestampDecoder), new _PropertyDecoder2.default(0x05, 'ticketEndTime').setDecoder(_helpers.timestampDecoder)];

    _this.parse(data, properties, config);
    return _this;
  }

  return ParkingTicketResponse;
}(_PropertyResponse3.default);

ParkingTicketResponse.identifier = [0x00, 0x47];
exports.default = ParkingTicketResponse;
