// hmkit/lib/Responses/Response.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _ChargingResponse = require('./ChargingResponse');

var _ChargingResponse2 = _interopRequireDefault(_ChargingResponse);

var _ChassisSettingsResponse = require('./ChassisSettingsResponse');

var _ChassisSettingsResponse2 = _interopRequireDefault(_ChassisSettingsResponse);

var _ClimateResponse = require('./ClimateResponse');

var _ClimateResponse2 = _interopRequireDefault(_ClimateResponse);

var _CruiseControlResponse = require('./CruiseControlResponse');

var _CruiseControlResponse2 = _interopRequireDefault(_CruiseControlResponse);

var _DashboardLightsResponse = require('./DashboardLightsResponse');

var _DashboardLightsResponse2 = _interopRequireDefault(_DashboardLightsResponse);

var _DiagnosticsResponse = require('./DiagnosticsResponse');

var _DiagnosticsResponse2 = _interopRequireDefault(_DiagnosticsResponse);

var _DoorLocksResponse = require('./DoorLocksResponse');

var _DoorLocksResponse2 = _interopRequireDefault(_DoorLocksResponse);

var _EngineResponse = require('./EngineResponse');

var _EngineResponse2 = _interopRequireDefault(_EngineResponse);

var _EmptyResponse = require('./EmptyResponse');

var _EmptyResponse2 = _interopRequireDefault(_EmptyResponse);

var _FailureMessageResponse = require('./FailureMessageResponse');

var _FailureMessageResponse2 = _interopRequireDefault(_FailureMessageResponse);

var _FirmwareVersionResponse = require('./FirmwareVersionResponse');

var _FirmwareVersionResponse2 = _interopRequireDefault(_FirmwareVersionResponse);

var _FuelingResponse = require('./FuelingResponse');

var _FuelingResponse2 = _interopRequireDefault(_FuelingResponse);

var _HistoricalResponse = require('./HistoricalResponse');

var _HistoricalResponse2 = _interopRequireDefault(_HistoricalResponse);

var _HomeChargerResponse = require('./HomeChargerResponse');

var _HomeChargerResponse2 = _interopRequireDefault(_HomeChargerResponse);

var _HonkHornFlashLightsResponse = require('./HonkHornFlashLightsResponse');

var _HonkHornFlashLightsResponse2 = _interopRequireDefault(_HonkHornFlashLightsResponse);

var _HoodResponse = require('./HoodResponse');

var _HoodResponse2 = _interopRequireDefault(_HoodResponse);

var _LightConditionsResponse = require('./LightConditionsResponse');

var _LightConditionsResponse2 = _interopRequireDefault(_LightConditionsResponse);

var _LightsResponse = require('./LightsResponse');

var _LightsResponse2 = _interopRequireDefault(_LightsResponse);

var _MaintenanceResponse = require('./MaintenanceResponse');

var _MaintenanceResponse2 = _interopRequireDefault(_MaintenanceResponse);

var _MobileResponse = require('./MobileResponse');

var _MobileResponse2 = _interopRequireDefault(_MobileResponse);

var _NaviDestinationResponse = require('./NaviDestinationResponse');

var _NaviDestinationResponse2 = _interopRequireDefault(_NaviDestinationResponse);

var _OffroadResponse = require('./OffroadResponse');

var _OffroadResponse2 = _interopRequireDefault(_OffroadResponse);

var _ParkingBrakeResponse = require('./ParkingBrakeResponse');

var _ParkingBrakeResponse2 = _interopRequireDefault(_ParkingBrakeResponse);

var _ParkingTicketResponse = require('./ParkingTicketResponse');

var _ParkingTicketResponse2 = _interopRequireDefault(_ParkingTicketResponse);

var _PowerTakeOffResponse = require('./PowerTakeOffResponse');

var _PowerTakeOffResponse2 = _interopRequireDefault(_PowerTakeOffResponse);

var _RaceResponse = require('./RaceResponse');

var _RaceResponse2 = _interopRequireDefault(_RaceResponse);

var _RooftopControlResponse = require('./RooftopControlResponse');

var _RooftopControlResponse2 = _interopRequireDefault(_RooftopControlResponse);

var _SeatsResponse = require('./SeatsResponse');

var _SeatsResponse2 = _interopRequireDefault(_SeatsResponse);

var _StartStopResponse = require('./StartStopResponse');

var _StartStopResponse2 = _interopRequireDefault(_StartStopResponse);

var _TachographResponse = require('./TachographResponse');

var _TachographResponse2 = _interopRequireDefault(_TachographResponse);

var _TheftAlarmResponse = require('./TheftAlarmResponse');

var _TheftAlarmResponse2 = _interopRequireDefault(_TheftAlarmResponse);

var _UsageResponse = require('./UsageResponse');

var _UsageResponse2 = _interopRequireDefault(_UsageResponse);

var _TrunkAccessResponse = require('./TrunkAccessResponse');

var _TrunkAccessResponse2 = _interopRequireDefault(_TrunkAccessResponse);

var _ValetModeResponse = require('./ValetModeResponse');

var _ValetModeResponse2 = _interopRequireDefault(_ValetModeResponse);

var _VehicleLocationResponse = require('./VehicleLocationResponse');

var _VehicleLocationResponse2 = _interopRequireDefault(_VehicleLocationResponse);

var _VehicleTimeResponse = require('./VehicleTimeResponse');

var _VehicleTimeResponse2 = _interopRequireDefault(_VehicleTimeResponse);

var _WeatherConditionsResponse = require('./WeatherConditionsResponse');

var _WeatherConditionsResponse2 = _interopRequireDefault(_WeatherConditionsResponse);

var _WiFiResponse = require('./WiFiResponse');

var _WiFiResponse2 = _interopRequireDefault(_WiFiResponse);

var _WindowsResponse = require('./WindowsResponse');

var _WindowsResponse2 = _interopRequireDefault(_WindowsResponse);

var _WindscreenResponse = require('./WindscreenResponse');

var _WindscreenResponse2 = _interopRequireDefault(_WindscreenResponse);

var _CapabilitiesResponse = require('./CapabilitiesResponse');

var _CapabilitiesResponse2 = _interopRequireDefault(_CapabilitiesResponse);

var _VehicleStatusResponse = require('./VehicleStatusResponse');

var _VehicleStatusResponse2 = _interopRequireDefault(_VehicleStatusResponse);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Response = function () {
  function Response(data) {
    var config = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, Response);

    this.parse = function () {
      var parsedValue = null;

      return function parse() {
        if (parsedValue !== null) return parsedValue;

        var bytes = this.bytes();
        var Parser = this.findParser(bytes);

        if (!Parser) {
          if (bytes.length === 0) return new _EmptyResponse2.default();
          return bytes;
        }

        parsedValue = new Parser(bytes, this.config);
        return parsedValue;
      };
    }();

    this.rawData = data;
    this.config = config;

    this.checkRawDataLength();

    this.parsers = [_CapabilitiesResponse2.default, _ChargingResponse2.default, _ChassisSettingsResponse2.default, _ClimateResponse2.default, _CruiseControlResponse2.default, _DashboardLightsResponse2.default, _DiagnosticsResponse2.default, _DoorLocksResponse2.default, _EngineResponse2.default, _FailureMessageResponse2.default, _FirmwareVersionResponse2.default, _FuelingResponse2.default, _HistoricalResponse2.default, _HomeChargerResponse2.default, _HonkHornFlashLightsResponse2.default, _HoodResponse2.default, _LightConditionsResponse2.default, _LightsResponse2.default, _MaintenanceResponse2.default, _MobileResponse2.default, _NaviDestinationResponse2.default, _OffroadResponse2.default, _ParkingBrakeResponse2.default, _ParkingTicketResponse2.default, _PowerTakeOffResponse2.default, _RaceResponse2.default, _RooftopControlResponse2.default, _SeatsResponse2.default, _StartStopResponse2.default, _TachographResponse2.default, _TheftAlarmResponse2.default, _UsageResponse2.default, _TrunkAccessResponse2.default, _ValetModeResponse2.default, _VehicleLocationResponse2.default, _VehicleStatusResponse2.default, _VehicleTimeResponse2.default, _WeatherConditionsResponse2.default, _WiFiResponse2.default, _WindowsResponse2.default, _WindscreenResponse2.default];
  }

  _createClass(Response, [{
    key: 'checkRawDataLength',
    value: function checkRawDataLength() {
      if (this.rawData.length < 2 && this.rawData.length !== 0) {
        throw new Error('Response string length invalid (length: ' + this.rawData.length + ' chars).');
      }
    }
  }, {
    key: 'bytes',
    value: function bytes() {
      return this.rawData;
    }
  }, {
    key: 'findParser',
    value: function findParser(bytes) {
      if (bytes.length === 0) {
        return null;
      }

      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = this.parsers[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var parser = _step.value;

          if (parser.identifier[0] === bytes[0] && parser.identifier[1] === bytes[1]) {
            return parser;
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      return null;
    }
  }]);

  return Response;
}();

exports.default = Response;
