// hmkit/lib/Responses/SeatsResponse.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _PropertyResponse2 = require('../PropertyResponse');

var _PropertyResponse3 = _interopRequireDefault(_PropertyResponse2);

var _PropertyDecoder = require('../PropertyDecoder');

var _PropertyDecoder2 = _interopRequireDefault(_PropertyDecoder);

var _OptionalPropertyDecoder = require('../OptionalPropertyDecoder');

var _OptionalPropertyDecoder2 = _interopRequireDefault(_OptionalPropertyDecoder);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SeatsResponse = function (_PropertyResponse) {
  _inherits(SeatsResponse, _PropertyResponse);

  /**
   * @property {Array} personsDetected (array) Persons detected ([{ seatPosition: (string: 'front_left|front_right|rear_right|rear_left|rear_center'), personDetected: (string: 'not_detected|detected')}])
   * @property {Array} seatbeltsFastened (array) Seatbelts fastened ([{ seatPosition: (string: 'front_left|front_right|rear_right|rear_left|rear_center', seatbeltFastened: (string: 'not_fastened|fastened')) }])
   *
   * @example SeatsResponse
    {
      personsDetected: [{
        value: {
          seatPosition: 'front_left',
          personDetected: 'not_detected'
        },
      }, {
        value: {
          seatPosition: 'front_right',
          personDetected: 'not_detected'
        },
      }, {
        value: {
          seatPosition: 'rear_right',
          personDetected: 'not_detected'
        },
      }, {
        value: {
          seatPosition: 'rear_left',
          personDetected: 'not_detected'
        }
      }],
      seatbeltsFastened: [{
        value: {
          seatPosition: 'front_left',
          seatbeltFastened: 'not_fastened',
        },
      }, {
        value: {
          seatPosition: 'front_right',
          seatbeltFastened: 'not_fastened',
        },
      }, {
        value: {
          seatPosition: 'rear_right',
          seatbeltFastened: 'not_fastened',
        },
      }, {
        value: {
          seatPosition: 'rear_left',
          seatbeltFastened: 'not_fastened'
        },
      }],
    }
   */
  function SeatsResponse(data, config) {
    _classCallCheck(this, SeatsResponse);

    var _this = _possibleConstructorReturn(this, (SeatsResponse.__proto__ || Object.getPrototypeOf(SeatsResponse)).call(this));

    var properties = [new _PropertyDecoder2.default(0x02, 'personsDetected').setOptionalSubProperties('seatPosition', [new _OptionalPropertyDecoder2.default(0x00, 'front_left').setDecoder(_this.personDetectedDecoder), new _OptionalPropertyDecoder2.default(0x01, 'front_right').setDecoder(_this.personDetectedDecoder), new _OptionalPropertyDecoder2.default(0x02, 'rear_right').setDecoder(_this.personDetectedDecoder), new _OptionalPropertyDecoder2.default(0x03, 'rear_left').setDecoder(_this.personDetectedDecoder), new _OptionalPropertyDecoder2.default(0x04, 'rear_center').setDecoder(_this.personDetectedDecoder)]), new _PropertyDecoder2.default(0x03, 'seatbeltsFastened').setOptionalSubProperties('seatPosition', [new _OptionalPropertyDecoder2.default(0x00, 'front_left').setDecoder(_this.seatbeltFastenedDecoder), new _OptionalPropertyDecoder2.default(0x01, 'front_right').setDecoder(_this.seatbeltFastenedDecoder), new _OptionalPropertyDecoder2.default(0x02, 'rear_right').setDecoder(_this.seatbeltFastenedDecoder), new _OptionalPropertyDecoder2.default(0x03, 'rear_left').setDecoder(_this.seatbeltFastenedDecoder), new _OptionalPropertyDecoder2.default(0x04, 'rear_center').setDecoder(_this.seatbeltFastenedDecoder)])];

    _this.parse(data, properties, config);
    return _this;
  }

  _createClass(SeatsResponse, [{
    key: 'personDetectedDecoder',
    value: function personDetectedDecoder(data) {
      return {
        personDetected: data[0] === 0x00 ? 'not_detected' : 'detected'
      };
    }
  }, {
    key: 'seatbeltFastenedDecoder',
    value: function seatbeltFastenedDecoder(data) {
      return {
        seatbeltFastened: data[0] === 0x00 ? 'not_fastened' : 'fastened'
      };
    }
  }]);

  return SeatsResponse;
}(_PropertyResponse3.default);

SeatsResponse.identifier = [0x00, 0x56];
exports.default = SeatsResponse;
