// hmkit/lib/Responses/SeatsResponse.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _PropertyResponse2 = require('../PropertyResponse');

var _PropertyResponse3 = _interopRequireDefault(_PropertyResponse2);

var _Property = require('../Property');

var _Property2 = _interopRequireDefault(_Property);

var _OptionalProperty = require('../OptionalProperty');

var _OptionalProperty2 = _interopRequireDefault(_OptionalProperty);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SeatsResponse = function (_PropertyResponse) {
  _inherits(SeatsResponse, _PropertyResponse);

  /**
   * @property {Array} personsDetected (array) Persons detected ([{ seatPosition: (string: 'front_left|front_right|rear_right|rear_left|rear_center'), personDetected: (string: 'not_detected|detected')}])
   * @property {Array} seatbeltsFastened (array) Seatbelts fastened ([{ seatPosition: (string: 'front_left|front_right|rear_right|rear_left|rear_center', seatbeltFastened: (string: 'not_fastened|fastened')) }])
   *
   * @example SeatsResponse {
      personsDetected: [{
        seatPosition: 'front_left',
        personDetected: 'detected'
      }, {
        seatPosition: 'front_right',
        personDetected: 'detected'
      }, {
        seatPosition: 'rear_right',
        personDetected: 'detected'
      }, {
        seatPosition: 'rear_left',
        personDetected: 'detected'
      }, {
        seatPosition: 'rear_center',
        personDetected: 'not_detected'
      }],
      seatbeltsFastened: [{
        seatPosition: 'front_left',
        seatbeltFastened: 'fastened'
      }, {
        seatPosition: 'front_right',
        seatbeltFastened: 'fastened'
      }, {
        seatPosition: 'rear_right',
        seatbeltFastened: 'fastened'
      }, {
        seatPosition: 'rear_left',
        seatbeltFastened: 'fastened'
      }, {
        seatPosition: 'rear_center',
        seatbeltFastened: 'not_fastened'
      }]
    }
    */
  function SeatsResponse(data) {
    _classCallCheck(this, SeatsResponse);

    var _this = _possibleConstructorReturn(this, (SeatsResponse.__proto__ || Object.getPrototypeOf(SeatsResponse)).call(this));

    var properties = [new _Property2.default(0x02, 'personsDetected').setOptionalSubProperties('seatPosition', [new _OptionalProperty2.default(0x00, 'front_left').setDecoder(_this.personDetectedDecoder), new _OptionalProperty2.default(0x01, 'front_right').setDecoder(_this.personDetectedDecoder), new _OptionalProperty2.default(0x02, 'rear_right').setDecoder(_this.personDetectedDecoder), new _OptionalProperty2.default(0x03, 'rear_left').setDecoder(_this.personDetectedDecoder), new _OptionalProperty2.default(0x04, 'rear_center').setDecoder(_this.personDetectedDecoder)]), new _Property2.default(0x03, 'seatbeltsFastened').setOptionalSubProperties('seatPosition', [new _OptionalProperty2.default(0x00, 'front_left').setDecoder(_this.seatbeltFastenedDecoder), new _OptionalProperty2.default(0x01, 'front_right').setDecoder(_this.seatbeltFastenedDecoder), new _OptionalProperty2.default(0x02, 'rear_right').setDecoder(_this.seatbeltFastenedDecoder), new _OptionalProperty2.default(0x03, 'rear_left').setDecoder(_this.seatbeltFastenedDecoder), new _OptionalProperty2.default(0x04, 'rear_center').setDecoder(_this.seatbeltFastenedDecoder)])];

    _this.parse(data, properties);
    return _this;
  }

  _createClass(SeatsResponse, [{
    key: 'personDetectedDecoder',
    value: function personDetectedDecoder(data) {
      return {
        personDetected: data[0] === 0x00 ? 'not_detected' : 'detected'
      };
    }
  }, {
    key: 'seatbeltFastenedDecoder',
    value: function seatbeltFastenedDecoder(data) {
      return {
        seatbeltFastened: data[0] === 0x00 ? 'not_fastened' : 'fastened'
      };
    }
  }]);

  return SeatsResponse;
}(_PropertyResponse3.default);

SeatsResponse.identifier = [0x00, 0x56];
exports.default = SeatsResponse;
