// hmkit/es/Commands/FuelingCommand.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import Command from './Command';
import BaseCommand from './BaseCommand';
import { validate, Joi } from '../validate';

var FuelingCommand = function (_BaseCommand) {
  _inherits(FuelingCommand, _BaseCommand);

  function FuelingCommand() {
    _classCallCheck(this, FuelingCommand);

    return _possibleConstructorReturn(this, (FuelingCommand.__proto__ || Object.getPrototypeOf(FuelingCommand)).apply(this, arguments));
  }

  _createClass(FuelingCommand, null, [{
    key: 'getState',

    /**
     * @function getState
     */
    value: function getState() {
      return new Command([0x00, 0x40, 0x00]);
    }

    /**
     * @function controlGasFlap
     *
     * @property {string} lock (string: 'unlocked|locked') Lock or unlock gas flap
     * @property {string} position (string: 'closed|opened|intermediate') Close or open gas flap
     */

  }, {
    key: 'controlGasFlap',
    value: function controlGasFlap(lock, position) {
      var lockBytes = [];
      var positionBytes = [];

      if (!!lock && lock.length > 0) {
        var lockStates = {
          unlocked: 0x00,
          locked: 0x01
        };

        validate([{
          value: lock,
          name: 'Lock',
          condition: Joi.string().required().valid(Object.keys(lockStates))
        }]);

        lockBytes = this.buildProperty(0x02, lockStates[lock]);
      }

      if (!!position && position.length > 0) {
        var positions = {
          closed: 0x00,
          opened: 0x01,
          intermediate: 0x02
        };

        validate([{
          value: position,
          name: 'Position',
          condition: Joi.string().required().valid(Object.keys(positions))
        }]);

        positionBytes = this.buildProperty(0x03, positions[position]);
      }

      return new Command([0x00, 0x40, 0x12].concat(_toConsumableArray(lockBytes), _toConsumableArray(positionBytes)));
    }
  }]);

  return FuelingCommand;
}(BaseCommand);

export default FuelingCommand;
