// hmkit/es/Commands/TrunkAccessCommand.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import Command from './Command';
import BaseCommand from './BaseCommand';

var TrunkAccessCommand = function (_BaseCommand) {
  _inherits(TrunkAccessCommand, _BaseCommand);

  function TrunkAccessCommand() {
    _classCallCheck(this, TrunkAccessCommand);

    return _possibleConstructorReturn(this, (TrunkAccessCommand.__proto__ || Object.getPrototypeOf(TrunkAccessCommand)).apply(this, arguments));
  }

  _createClass(TrunkAccessCommand, null, [{
    key: 'getState',

    /**
     * @function getState
     */
    value: function getState() {
      return new Command([0x00, 0x21, 0x00]);
    }

    /**
     * @function unlock
     */

  }, {
    key: 'unlock',
    value: function unlock() {
      return new Command([0x00, 0x21, 0x12].concat(_toConsumableArray(this.buildProperty(0x01, 0x00))));
    }

    /**
     * @function lock
     */

  }, {
    key: 'lock',
    value: function lock() {
      return new Command([0x00, 0x21, 0x12].concat(_toConsumableArray(this.buildProperty(0x01, 0x01))));
    }

    /**
     * @function open
     */

  }, {
    key: 'open',
    value: function open() {
      return new Command([0x00, 0x21, 0x12].concat(_toConsumableArray(this.buildProperty(0x02, 0x01))));
    }

    /**
     * @function close
     */

  }, {
    key: 'close',
    value: function close() {
      return new Command([0x00, 0x21, 0x12].concat(_toConsumableArray(this.buildProperty(0x02, 0x00))));
    }

    /**
     * @function controlTrunk
     *
     * @property {Boolean} isLocked (boolean) Lock or unlock trunk
     * @property {Boolean} isOpen (boolean) Open or close trunk
     */

  }, {
    key: 'controlTrunk',
    value: function controlTrunk(isLocked, isOpen) {
      return new Command([0x00, 0x21, 0x12].concat(_toConsumableArray(this.buildProperty(0x01, isLocked ? 0x01 : 0x00)), _toConsumableArray(this.buildProperty(0x02, isOpen ? 0x01 : 0x00))));
    }
  }]);

  return TrunkAccessCommand;
}(BaseCommand);

export default TrunkAccessCommand;
