// hmkit/es/Responses/ClimateResponse.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import PropertyResponse from '../PropertyResponse';
import Property from '../Property';
import OptionalProperty from '../OptionalProperty';
import { activeInactiveDecoder, getRoundedIeee754ToBase10 } from '../helpers';

var ClimateResponse = function (_PropertyResponse) {
  _inherits(ClimateResponse, _PropertyResponse);

  /**
   * @property {Number} insideTemperature (number) Temperature in celsius formatted in 4-bytes per IEEE 754
   * @property {Number} outsideTemperature (number) Temperature in celsius formatted in 4-bytes per IEEE 754
   * @property {Number} driverTemperatureSetting (number) Temperature in celsius formatted in 4-bytes per IEEE 754
   * @property {Number} passengerTemperatureSetting (number) Temperature in celsius formatted in 4-bytes per IEEE 754
   * @property {String} hvacState (string) HVAC state
   * @property {String} defoggingState (string) Defogging state
   * @property {String} defrostingState (string) Defrosting state
   * @property {String} ionisingState (string) Ionising state
   * @property {Number} defrostingTemperature (number) Temperature in celsius formatted in 4-bytes per IEEE 754
   * @property {Object} hvacWeekdayStartingTimes (array) Auto-HVAC activated on days ([{ weekday: (string: 'monday|tuesday|wednesday|thursday|friday|saturday|sunday|automatic', hour: (number), minute: (number) )}])
   * @property {Number} rearTemperature (number) Temperature in celsius formatted in 4-bytes per IEEE 754
   *
   * @example ClimateResponse
    {
      insideTemperature: 23,
      outsideTemperature: 18,
      driverTemperatureSetting: 23,
      passengerTemperatureSetting: 22,
      hvacState: 'inactive',
      defoggingState: 'inactive',
      defrostingState: 'inactive',
      ionisingState: 'inactive',
      defrostingTemperature: 23,
      hvacWeekdayStartingTimes: [{
        weekday: 'monday',
        hour: 8,
        minute: 0
      }, {
        weekday: 'tuesday',
        hour: 8,
        minute: 0
      }, {
        weekday: 'wednesday',
        hour: 8,
        minute: 0
      }, {
        weekday: 'thursday',
        hour: 8,
        minute: 0
      }, {
        weekday: 'friday',
        hour: 8,
        minute: 0
      }, {
        weekday: 'saturday',
        hour: 8,
        minute: 0
      }, {
        weekday: 'sunday',
        hour: 8,
        minute: 0
      }],
      rearTemperatureSetting: 22
    }
   */
  function ClimateResponse(data) {
    _classCallCheck(this, ClimateResponse);

    var _this = _possibleConstructorReturn(this, (ClimateResponse.__proto__ || Object.getPrototypeOf(ClimateResponse)).call(this));

    var properties = [new Property(0x01, 'insideTemperature').setDecoder(getRoundedIeee754ToBase10(2)), new Property(0x02, 'outsideTemperature').setDecoder(getRoundedIeee754ToBase10(2)), new Property(0x03, 'driverTemperatureSetting').setDecoder(getRoundedIeee754ToBase10(2)), new Property(0x04, 'passengerTemperatureSetting').setDecoder(getRoundedIeee754ToBase10(2)), new Property(0x05, 'hvacState').setDecoder(activeInactiveDecoder()), new Property(0x06, 'defoggingState').setDecoder(activeInactiveDecoder()), new Property(0x07, 'defrostingState').setDecoder(activeInactiveDecoder()), new Property(0x08, 'ionisingState').setDecoder(activeInactiveDecoder()), new Property(0x09, 'defrostingTemperature').setDecoder(getRoundedIeee754ToBase10(2)), new Property(0x0b, 'hvacWeekdayStartingTimes').setOptionalSubProperties('weekday', [new OptionalProperty(0x00, 'monday').setDecoder(_this.weekdayDecoder), new OptionalProperty(0x01, 'tuesday').setDecoder(_this.weekdayDecoder), new OptionalProperty(0x02, 'wednesday').setDecoder(_this.weekdayDecoder), new OptionalProperty(0x03, 'thursday').setDecoder(_this.weekdayDecoder), new OptionalProperty(0x04, 'friday').setDecoder(_this.weekdayDecoder), new OptionalProperty(0x05, 'saturday').setDecoder(_this.weekdayDecoder), new OptionalProperty(0x06, 'sunday').setDecoder(_this.weekdayDecoder), new OptionalProperty(0x07, 'automatic').setDecoder(_this.weekdayDecoder)]), new Property(0x0c, 'rearTemperatureSetting').setDecoder(getRoundedIeee754ToBase10(2))];

    _this.parse(data, properties);
    return _this;
  }

  _createClass(ClimateResponse, [{
    key: 'weekdayDecoder',
    value: function weekdayDecoder(data) {
      return {
        hour: data[0],
        minute: data[1]
      };
    }
  }]);

  return ClimateResponse;
}(PropertyResponse);

ClimateResponse.identifier = [0x00, 0x24];
export default ClimateResponse;
