// hmkit/es/Responses/MaintenanceResponse.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import PropertyResponse from '../PropertyResponse';
import PropertyDecoder from '../PropertyDecoder';
import { bytesSum, switchDecoder, timestampDecoder } from '../helpers';
import { bytesToString } from '../encoding';

var MaintenanceResponse = function (_PropertyResponse) {
  _inherits(MaintenanceResponse, _PropertyResponse);

  /**
   * @property {Number} daysToNextService (number) Number of days until next servicing of the car, whereas negative is overdue
   * @property {Number} kilometersToNextService (number) Amount of kilometers until next servicing of the car
   * @property {Number} cbsReportsCount (number) The number of CBS reports
   * @property {Number} monthsToExhaustInspection (number) Number of months until exhaust inspection
   * @property {String} teleserviceAvailability (string: 'pending|idle|successful|error') Teleservice availability
   * @property {Number} serviceDistanceThreshold (number) Distance threshold for Service
   * @property {Number} serviceTimeThreshold (number) Time threshold, in weeks, for Service
   * @property {Date} automaticTeleserviceCallDate (date) Automatic teleservice call date
   * @property {Date} teleserviceBatteryCallDate (date) Teleservice battery call date
   * @property {Date} nextInspectionDate (date) Next inspection date
   * @property {Array} conditionBasedServices (Array) Condition based services ([{ year: (number), month: (number), cbsIdentifier: (number), dueStatus: (string: 'ok|pending|overdue'), cbsText: (string), description: (string) }])
   * @property {Date} brakeFluidChangeDate (Date) Brake fluid change date
   *
   * @example MaintenanceResponse
    {
      daysToNextService: {
        value: 400,
      },
      kilometersToNextService: {
        value: 30000,
      },
      cbsReportsCount: {
        value: 0,
      },
      monthsToExhaustInspection: {
        value: 0,
      },
      teleserviceAvailability: {
        value: 'pending',
      },
      serviceDistanceThreshold: {
        value: 0,
      },
      serviceTimeThreshold: {
        value: 0,
      },
      automaticTeleserviceCallDate: {
        value: '2018-10-22T12:10:33.769Z',
      },
      teleserviceBatteryCallDate: {
        value: '2018-10-22T12:10:33.769Z',
      },
      nextInspectionDate: {
        value: '2018-10-22T12:10:33.769Z',
      },
      conditionBasedServices: [{
        value: {
          year: 2018,
          month: 8,
          cbsIdentifier: 123,
          dueStatus: 'ok',
          cbsText: '',
          description: '',
        },
      }],
      brakeFluidChangeDate: {
        value: '2018-10-22T12:10:33.769Z',
      },
    }
   */
  function MaintenanceResponse(data, config) {
    _classCallCheck(this, MaintenanceResponse);

    var _this = _possibleConstructorReturn(this, (MaintenanceResponse.__proto__ || Object.getPrototypeOf(MaintenanceResponse)).call(this));

    var properties = [new PropertyDecoder(0x01, 'daysToNextService').setDecoder(bytesSum), new PropertyDecoder(0x02, 'kilometersToNextService').setDecoder(bytesSum), new PropertyDecoder(0x03, 'cbsReportsCount').setDecoder(bytesSum), new PropertyDecoder(0x04, 'monthsToExhaustInspection').setDecoder(bytesSum), new PropertyDecoder(0x05, 'teleserviceAvailability').setDecoder(switchDecoder({
      0x00: 'pending',
      0x01: 'idle',
      0x02: 'succesful',
      0x03: 'error'
    })), new PropertyDecoder(0x06, 'serviceDistanceThreshold').setDecoder(bytesSum), new PropertyDecoder(0x07, 'serviceTimeThreshold').setDecoder(bytesSum), new PropertyDecoder(0x08, 'automaticTeleserviceCallDate').setDecoder(timestampDecoder), new PropertyDecoder(0x09, 'teleserviceBatteryCallDate').setDecoder(timestampDecoder), new PropertyDecoder(0x0a, 'nextInspectionDate').setDecoder(timestampDecoder), new PropertyDecoder(0x0b, 'conditionBasedServices').setDecoder(_this.conditionBasedServicesDecoder).array(), new PropertyDecoder(0x0c, 'brakeFluidChangeDate').setDecoder(timestampDecoder)];

    _this.parse(data, properties, config);
    return _this;
  }

  _createClass(MaintenanceResponse, [{
    key: 'conditionBasedServicesDecoder',
    value: function conditionBasedServicesDecoder(bytes) {
      var cbsTextLength = bytesSum(bytes.slice(5, 7));
      var cbsText = bytesToString(bytes.slice(7, 7 + cbsTextLength));
      var descriptionLength = bytesSum(bytes.slice(7 + cbsTextLength, 9 + cbsTextLength));
      var description = bytesToString(bytes.slice(9 + cbsTextLength, 9 + cbsTextLength + descriptionLength));

      return {
        year: bytes[0] + 2000,
        month: bytes[1],
        cbsIdentifier: bytesSum(bytes.slice(2, 4)),
        dueStatus: switchDecoder({
          0x00: 'ok',
          0x01: 'pending',
          0x02: 'overdue'
        })([bytes[4]]),
        cbsText: cbsText,
        description: description
      };
    }
  }]);

  return MaintenanceResponse;
}(PropertyResponse);

MaintenanceResponse.identifier = [0x00, 0x34];
export default MaintenanceResponse;
