// hmkit/es/Responses/UsageResponse.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import PropertyResponse from '../PropertyResponse';
import Property from '../Property';
import OptionalProperty from '../OptionalProperty';
import { progressDecoder, bytesSum, getRoundedIeee754ToBase10, dateDecoder } from '../helpers';

var UsageResponse = function (_PropertyResponse) {
  _inherits(UsageResponse, _PropertyResponse);

  /**
   * @property {Number} averageWeeklyDistance (number) Average weekly distance in km
   * @property {Number} averageWeeklyDistanceLongRun (number) Average weekyl distance, over long term, in km
   * @property {Number} accelerationEvaluation (number) Acceleration evaluation percentage (0..1)
   * @property {Number} drivingStyleEvaluation (number) Driving style's evaluation percentage (0..1)
   * @property {Array} drivingModesActivationPeriods (array) Driving modes activation periods ([{ drivingMode: (string: 'regular|eco|sport|sport_plus|eco_plus'), period: (number) }])
   * @property {Array} drivingModesEnergyConsumptions (array) Driving modes energy consumptions ([{ drivingMode: (string: 'regular|eco|sport|sport_plus|eco_plus'), consumption: (number) }])
   * @property {Number} lastTripEnergyConsumption (number) Energy consumption in the last trip in kWh
   * @property {Number} lastTripFuelConsumption (number) Fuel consumption in the last trip in L
   * @property {Number} mileageAfterLastTrip (number) Mileage after the last trip in km
   * @property {Number} lastTripElectricPortion (number) Portion of the last trip used in electric mode
   * @property {Number} lastTripAverageEnergyRecuperation (number) Energy recuperation rate for last trip, in kWh / 100 km
   * @property {Number} lastTripBatteryRemaining (number) Battery % remaining after last trip (0..1)
   * @property {Date} lastTripDate (date) Last trip date
   * @property {Number} currentFuelConsumption (number) Current fuel consumption formatted in 4-bytes per IEEE 754
   * @property {Number} averageFuelConsumption (number) Average fuel consumption formatted in 4-bytes per IEEE 754
   *
   * @example UsageResponse
    {
      currentFuelConsumption: 8.7,
      averageFuelConsumption: 6.2,
    }
   */

  function UsageResponse(data) {
    _classCallCheck(this, UsageResponse);

    var _this = _possibleConstructorReturn(this, (UsageResponse.__proto__ || Object.getPrototypeOf(UsageResponse)).call(this));

    var properties = [new Property(0x01, 'averageWeeklyDistance').setDecoder(bytesSum), new Property(0x02, 'averageWeeklyDistanceLongRun').setDecoder(bytesSum), new Property(0x03, 'accelerationEvaluation').setDecoder(progressDecoder), new Property(0x04, 'drivingStyleEvaluation').setDecoder(progressDecoder), new Property(0x05, 'drivingModesActivationPeriods').setOptionalSubProperties('drivingMode', [new OptionalProperty(0x00, 'regular').setDecoder(_this.activationPeriodDecoder), new OptionalProperty(0x01, 'eco').setDecoder(_this.activationPeriodDecoder), new OptionalProperty(0x02, 'sport').setDecoder(_this.activationPeriodDecoder), new OptionalProperty(0x03, 'sport_plus').setDecoder(_this.activationPeriodDecoder), new OptionalProperty(0x04, 'eco_plus').setDecoder(_this.activationPeriodDecoder)]), new Property(0x06, 'drivingModesEnergyConsumptions').setOptionalSubProperties('drivingMode', [new OptionalProperty(0x00, 'regular').setDecoder(_this.energyConsumptionDecoder), new OptionalProperty(0x01, 'eco').setDecoder(_this.energyConsumptionDecoder), new OptionalProperty(0x02, 'sport').setDecoder(_this.energyConsumptionDecoder), new OptionalProperty(0x03, 'sport_plus').setDecoder(_this.energyConsumptionDecoder), new OptionalProperty(0x04, 'eco_plus').setDecoder(_this.energyConsumptionDecoder)]), new Property(0x07, 'lastTripEnergyConsumption').setDecoder(getRoundedIeee754ToBase10(1)), new Property(0x08, 'lastTripFuelConsumption').setDecoder(getRoundedIeee754ToBase10(1)), new Property(0x09, 'mileageAfterLastTrip').setDecoder(bytesSum), new Property(0x0a, 'lastTripElectricPortion').setDecoder(bytesSum), new Property(0x0b, 'lastTripAverageEnergyRecuperation').setDecoder(getRoundedIeee754ToBase10(2)), new Property(0x0c, 'lastTripBatteryRemaining').setDecoder(progressDecoder), new Property(0x0d, 'lastTripDate').setDecoder(dateDecoder), new Property(0x0e, 'averageFuelConsumption').setDecoder(getRoundedIeee754ToBase10(2)), new Property(0x0f, 'currentFuelConsumption').setDecoder(getRoundedIeee754ToBase10(2))];

    _this.parse(data, properties);
    return _this;
  }

  _createClass(UsageResponse, [{
    key: 'activationPeriodDecoder',
    value: function activationPeriodDecoder(bytes) {
      return {
        period: progressDecoder(bytes)
      };
    }
  }, {
    key: 'energyConsumptionDecoder',
    value: function energyConsumptionDecoder(bytes) {
      return {
        consumption: getRoundedIeee754ToBase10(1)(bytes)
      };
    }
  }]);

  return UsageResponse;
}(PropertyResponse);

UsageResponse.identifier = [0x00, 0x68];
export default UsageResponse;
